src/layers/GeomodelLayer.ts

```diff
@@ -0,0 +1,146 @@
+
+  curveOptions = { samples: 50, knot: 0.5 };
+
+  constructor(id: string, options: GeomodelLayerOptions) {
+    super(id, options);
+    this.options = {
+      ...options,
+    };
+    this.render = this.render.bind(this);
+  }
+
+  onUpdate(event: OnUpdateEvent) {
+    super.onUpdate(event);
+    this.render(event);
+  }
+
+  onRescale(event: OnRescaleEvent) {
+    super.onRescale(event);
+    this.render(event);
+  }
+
+  render(event: OnRescaleEvent | OnUpdateEvent) {
+    const { xscale, yscale, data } = event;
+    const scaledData = this.generateScaledData(data, xscale, yscale);
+
+    // for each layer
+    // paint from top down, with color
+    // if missing bottom data is null, use next area top line as bottom
+
+    data.forEach((s: GeoModelData, index: number) =>
+      this.drawArea(s, index, scaledData[index]),
+    );
+  }
+
+  drawArea = (
+    s: GeoModelData,
+    index: number,
+    scaledData: { top: [number, number][]; bottom: [number, number][] },
+  ): void => {
+    const area = new Graphics();
+    area.lineStyle(4, s.color, 1);
+    area.beginFill(s.color);
+    this.renderAreaTopLine(area, scaledData.top, scaledData.bottom);
+    area.endFill();
+    this.ctx.stage.addChild(area);
+  };
+
+  generateScaledData = (
+    data: GeoModelData[],
+    xscale: ScaleLinear<number, number>,
+    yscale: ScaleLinear<number, number>,
+  ): { top: [number, number][]; bottom: [number, number][] }[] => {
+    return data.map((stratLayer: GeoModelData, index: number) => {
+      const top = this.generateScaledLayer(
+        stratLayer.pos,
+        stratLayer.md,
+        xscale,
+        yscale,
+      );
+      let bottom: [number, number][] = null;
+      if (stratLayer.bottomPos == null || stratLayer.bottomMd == null) {
+        // last layer, draw to yscale bottom
+        if (index >= data.length - 1) {
+          bottom = [
+            [xscale.range()[0], yscale.range()[1]],
+            [xscale.range()[0], yscale.range()[1]],
+            [xscale.range()[1], yscale.range()[1]],
+            [xscale.range()[1], yscale.range()[1]],
+          ];
+        } else {
+          // use next area top line as bottom
+          bottom = this.generateScaledLayer(
+            data[index + 1].pos,
+            data[index + 1].md,
+            xscale,
+            yscale,
+          );
+        }
+      } else {
+        // use provided bottom data
+        bottom = this.generateScaledLayer(
+          stratLayer.bottomPos,
+          stratLayer.bottomMd,
+          xscale,
+          yscale,
+        );
+      }
+      return { top, bottom };
+    });
+  };
+
+  generateScaledLayer = (
+    pos: [number, number][],
+    md: number[],
+    xscale: ScaleLinear<number, number>,
+    yscale: ScaleLinear<number, number>,
+  ): [number, number][] => {
+    return pos.map((posPoint: [number, number], index) => [
+      xscale(posPoint[0]),
+      yscale(md[index]),
+    ]);
+  };
+
+  private renderAreaTopLine = (
```

Comment:
If I understand this correctly, it renders the area polygon and should be named accordingly

Reply:
What will a line surface be? One depth, or a usual surface with only a repeating depth? (And no bottom values? Without going until the next surface layer?)

Reply:
The surfaces which are rendered as lines is typically the top of a formation. In rep we also have oil/water contact which are the flat lines, which should maybe be added as another feature. Suggest we add surfacelines as an additional feature or atleast in a seperate pull request to get geolayers into master ASAP.